Add UCN documentation [CN-1131] (#190)
hazelcast/hazelcast-platform-operator-docs: docs/modules/ROOT/pages/user-code-namespaces.adoc

= User Code Namespaces
:description: User Code Namespaces provide a container for Java classpath resources, such as user code and accompanying artifacts like property files. This provides namespace isolation to ensure that access to resources in different namespaces can be managed through configuration. From Hazelcast Platform Operator, you can deploy your application code from external buckets using UserCodeNamespace CRs. 

{description}

See xref:hazelcast:clusters:user-code-namespaces.adoc[User Code Namespaces] for more detailed information about the configuration and behavior of User Code Namespaces in Hazelcast.

== Enabling User Code Namespaces

To enable User Code Namespaces you must configure the `userCodeNamespaces` section in the Hazelcast custom resource.

.Example of Hazelcast with User Code Namespaces enabled
[source,yaml,subs="attributes+"]
----
include::ROOT:example$/hazelcast-ucn.yaml[]
----

> == Creating a new User Code Namespace

Reviewer: ```suggestion
== Creating a User Code Namespace
```

> To create a new User Code Namespace you must create a new `UserCodeNamespace` CR. The name of the CR will be used as the name of the created User Code Namespace. The JAR files stored in the configured bucket will be uploaded to the Hazelcast cluster under the newly created namespace.

Reviewer: ```suggestion
To create a User Code Namespace, you must create a new `UserCodeNamespace` CR. The name of the CR is used as the name of the created User Code Namespace. The JAR files stored in the configured bucket are uploaded to the Hazelcast cluster under the newly created namespace.
```

> .Example of User Code Namespaces

Reviewer 2: Shall we add multiple tabs for different cloud providers (AWS, GCP, Azure) as we do in backup and restore docs?

[source,yaml,subs="attributes+"]
----
include::ROOT:example$/user-code-namespace.yaml[]
----

> <1> The bucket URI where backup data will be stored in

Reviewer: ```suggestion
<1> The bucket URI in which to store backup data
```

== Enabling User Code Namespaces

To enable User Code Namespaces you must configure the `userCodeNamespaces` section in the Hazelcast custom resource.

.Example of Hazelcast with User Code Namespaces enabled
[source,yaml,subs="attributes+"]
----
include::ROOT:example$/hazelcast-ucn.yaml[]
----

== Creating a User Code Namespace

To create a User Code Namespace, you must create a new `UserCodeNamespace` CR. The name of the CR is used as the name of the created User Code Namespace. The JAR files stored in the configured bucket are uploaded to the Hazelcast cluster under the newly created namespace.

[tabs]
====

AWS::
+
--
User Code Namespaces in AWS S3:
[source,yaml,subs="attributes+"]
----
include::ROOT:example$/user-code-namespace-aws.yaml[]
----
--

GCP::
+
--
User Code Namespaces in GCP:
[source,yaml,subs="attributes+"]
----
include::ROOT:example$/user-code-namespace-gcp.yaml[]
----
--

Azure::
+
--
User Code Namespaces in Azure Blob:
[source,yaml,subs="attributes+"]
----
include::ROOT:example$/user-code-namespace-azure.yaml[]
----
--
====

> <1> The bucket URI in which to store backup data

Reviewer: ```suggestion
<1> The bucket URI in which to store backup data.
```
